torch/csrc/jit/register_prim_ops.cpp

```diff
@@ -103,6 +103,70 @@ void checkImplicitTensorToNum(at::Tensor t, bool toInt) {
   }
 }
 
+// Convert the tensor pointed to by \p data to a nested list. \p dim is the
+// number of dimensions in the tensor and \p cur_dim is the dimension being
+// processed by the current invocation. \p ty is the expected output IR type of
+// the operation. \p sizes and \p strides are the sizes and strides of the
+// tensor operand and \p element_size is the size in bytes of one tensor
+// element.
+c10::impl::GenericList tensorToListRecursive(
+    char* data,
+    int64_t cur_dim,
+    int64_t dim,
+    TypePtr ty,
+    at::IntArrayRef sizes,
+    at::IntArrayRef strides,
+    size_t element_size) {
+  // If ty is a ListType, get the element type.
+  if (auto list_type = ty->cast<ListType>()) {
+    ty = list_type->getElementType();
+  }
+
+  // Make the result list consisting of elements of type ty. Since this
+  // invocation is processing dimension cur_dim, there will be sizes[cur_dim]
+  // output elements.
+  auto result = c10::impl::GenericList(ty);
+  result.reserve(sizes[cur_dim]);
+
+  if (auto list_type = ty->cast<ListType>()) {
+    // If ty is still a list type, then tensorToListRecursive needs to be called
+    // recursively on each slice of the tensor in the current dimension.
+    for (int64_t i = 0, e = sizes[cur_dim]; i < e; ++i) {
+      auto inner_result = tensorToListRecursive(
+          data, cur_dim + 1, dim, ty, sizes, strides, element_size);
+      result.emplace_back(inner_result);
+      data += strides[cur_dim] * element_size;
+    }
+  } else {
+    // If ty is not a list type, this invocation must be processing the last
+    // dimension, so sizes[cur_dim] consecutive scalars of the appropriate type
+    // should be read and stored in the result list.
+    if (ty == IntType::get()) {
+      for (int64_t i = 0, e = sizes[cur_dim]; i < e; ++i) {
+        int64_t scalar = *(int64_t*)data;
+        result.emplace_back(scalar);
+        data += element_size;
```

Comment:
I'm not 100% sure but I think this should be `data += strides[dim] * elementSize`. i'm think that's what the equivalent logic in `recursive_to_list` does: https://github.com/pytorch/pytorch/blob/f531815526c69f432e46fadece44f5d3a9b70e30/torch/csrc/utils/tensor_list.cpp#L26

Also: this isn't necessary, but it might make this code a little bit easier to reason about if it had a closer correspondence to `recursive_to_list`. 

`tensorToListRecursive` could return an `IValue` instead of `c10::impl::GenericList`. When `num_tensor_dim` == `cur_dim`, return the Scalar IValue, otherwise return the GenericList.  I think this would also make it so that you no longer need to specially handle 0-dim tensors in the body of the operator.

Reply:
Done and done.

I didn't realize that transposition of a tensor adjusts the strides/sizes rather than actually transposing the data. So yes, it should be `data += strides[dim] * elementSize` even if it's the last dimension. The unit tests I added for noncontiguous tensors exposed this issue. Thanks!